### ISSUE TITLE
K-means clustering: Klynge-personaer og kollaborativ filtering

### ISSUE THREAD

**Issue by andreas-t-hjertaker** [andreas-t-hjertaker/suksess#13]
## Beskrivelse
Implementer K-means clustering som grupperer elever med lignende profiler. Brukes for klynge-delt innhold, kollaborativ filtering, og admin-innsikt.

### Anbefalinger
- **k = 6-8 klynger** anbefalt basert på Silhouette-analyse av Big Five + RIASEC feature-vektorer
- **Feature-vektor (17 dimensjoner):**
  - Big Five: Openness, Conscientiousness, Extraversion, Agreeableness, Neuroticism (5)
  - RIASEC: Realistic, Investigative, Artistic, Social, Enterprising, Conventional (6)
  - Tillegg: interesser (3-5 kategorier), snittkarakter VGS, alder, kjønn (6)
- **Normalisering:** Z-score standardisering av alle features
- **Initialisering:** K-means++ for bedre konvergens

### Infrastruktur
- **Cloud Run (Python)** — batch-jobb, ikke persistent server
- scikit-learn `MiniBatchKMeans` for effektiv kjøring
- Nattlig Pub/Sub trigger → Cloud Run job
- Resultater tilbake til Firestore (`users/{uid}.clusterId`)

### Pipeline
```
Firestore (brukerprofiler) → Cloud Run Python →
  1. Hent alle profiler med komplett Big Five + RIASEC
  2. Normaliser features (StandardScaler)
  3. Kjør KMeans(n_clusters=k, init='k-means++')
  4. Evaluer: Silhouette Score, Calinski-Harabasz
  5. Generer klynge-sentroids
  6. Tilordne clusterId til brukerprofiler
  7. AI-generer klynge-persona (LLM-kall med sentroid-verdier)
  8. Lagre til Firestore
```

### Kollaborativ Filtering
- **Cluster-basert:** «Elever i din klynge valgte oftest: [X, Y, Z]»
- Beregn popularitet av studieprogram, karriereveier, og programfag per klynge
- Oppdater `cluster_recommendations` collection nattlig

### Evaluering
- Silhouette Score > 0.3 er akseptabelt for personlighetsdata
- Kjør med k = 5, 6, 7, 8 og velg beste score
- Visualiser med PCA/t-SNE for admin-dashboard

## Tasks
- [ ] Bygg Cloud Run Python-service: `suksess-clustering`
- [ ] Definer feature-vektor: Big Five (5) + RIASEC (6) + interesser + karaktersnitt = ~17 dimensjoner
- [ ] Implementer StandardScaler normalisering
- [ ] Implementer K-means++ med k=6-8, evaluer med Silhouette Score
- [ ] Lagre `clusterId` på brukerprofil i Firestore
- [ ] Generer klynge-beskrivelser med LLM (persona per klynge)
- [ ] Bygg kollaborativ filtering: popularitet av valg per klynge
- [ ] Implementer nattlig Pub/Sub trigger → Cloud Run job
- [ ] Bygg PCA/t-SNE klyngevisualisering for admin-dashboard
- [ ] Anonym klyngefordeling (ingen PII i admin-view)

## Akseptansekriterier
- Clustering kjører nattlig og oppdaterer brukerprofilene
- Minimum 6 distinkte klynger med AI-genererte persona-beskrivelser
- Silhouette Score > 0.3
- Kollaborativ filtering gir relevante anbefalinger
- Admin kan se anonym klyngefordeling

## Referanser
- [scikit-learn KMeans](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html)
- [Cloud Run Jobs](https://cloud.google.com/run/docs/create-jobs)
- Fullstendig research: `/research/weaviate-kmeans.md`
